### PR TITLE
Generated client: embed generated methods into VMIRS interface

### PR DESCRIPTION
**Pull request by fossedihelm**
### What this PR does
Before this PR:
We have our custom `virtualMachineInstanceReplicasetInterface` interface with the common `Get`, `Delete`, etc...
From https://github.com/kubevirt/kubevirt/pull/11297 the common methods implementation works as a wrapper with different signatures for the generated ones, redirecting the requests to them.

After this PR:
Convert using the generated client signatures instead of the custom one. The generated `virtualMachineInstanceReplicaset`  is embedded into our custom one, also providing new `DeleteCollection` and `Watch` methods.
The changes inherited are:
- `Patch` and `PatchStatus` now use metav1.PatchOptions{} instead of *metav1.PatchOptions{}.
- `List` now uses metav1.ListOptions{} instead of *metav1.ListOptions{}.
- `Get` now uses metav1.GetOptions{} instead of *metav1.GetOptions{}.
- `Delete` now uses metav1.DeleteOptions{} instead of *metav1.DeleteOptions{}.
- `Create` now receives metav1.CreateOptions{} too.
- `Update` and `UpdateStatus` now receive metav1.UpdateOptions{} too.
- Added `DeleteCollection`, `Watch` methods.
- All the methods above take a  `Context` as first paramether.

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Special notes for your reviewer
For an easier review process the commits was splitted in the following manner:
1. Align our custom common methods' signature to the generated one in the following order:
     - `Create` - introduced `CreateOptions` and introduced a `Context`
     - `Get` - used `GetOptions` instead of the pointer and introduced a `Context`
     - `List` - used `ListOptions` instead of the pointer and introduced a `Context`
     - `Update` - introduced `UpdateOptions` and introduced a `Context`
     - `Delete` - used `DeleteOptions` instead of the pointer and introduced a `Context`
     - `Patch` and `PatchStatus` - used `PatchOptions` instead of the pointer and introduced a `Context`
     - `UpdateStatus` - introduced `UpdateOptions` and introduced a `Context`
2. Embed the generated `virtualMachineInstanceReplicaset` into the custom one.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] ~Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required~
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] ~Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test~
- [ ] ~Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.~
- [ ] ~Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered~

### Release note

```release-note
NONE
```

### CODE REVIEW

File: staging/src/kubevirt.io/client-go/kubecli/replicaset.go [hunk 1]
```diff
@@ -87,8 +87,8 @@ func (v *rc) List(options k8smetav1.ListOptions) (replicasetList *v1.VirtualMach
 	return
 }
 
-func (v *rc) Create(replicaset *v1.VirtualMachineInstanceReplicaSet) (result *v1.VirtualMachineInstanceReplicaSet, err error) {
-	result, err = v.VirtualMachineInstanceReplicaSetInterface.Create(context.Background(), replicaset, k8smetav1.CreateOptions{})
+func (v *rc) Create(ctx context.Context, replicaset *v1.VirtualMachineInstanceReplicaSet, opts k8smetav1.CreateOptions) (result *v1.VirtualMachineInstanceReplicaSet, err error) {
```
Comment (orelmisan) [inline staging/src/kubevirt.io/client-go/kubecli/replicaset.go:89]: Please consider dropping the named return values.
Same for other methods.

File: staging/src/kubevirt.io/client-go/kubecli/kubevirt.go [hunk 2]
```diff
@@ -268,15 +268,9 @@ type VirtualMachineInstanceInterface interface {
...
 
 type ReplicaSetInterface interface {
-	Get(ctx context.Context, name string, options metav1.GetOptions) (*v1.VirtualMachineInstanceReplicaSet, error)
-	List(ctx context.Context, opts metav1.ListOptions) (*v1.VirtualMachineInstanceReplicaSetList, error)
-	Create(ctx context.Context, replicaset *v1.VirtualMachineInstanceReplicaSet, opts metav1.CreateOptions) (*v1.VirtualMachineInstanceReplicaSet, error)
-	Update(ctx context.Context, virtualMachineInstanceReplicaSet *v1.VirtualMachineInstanceReplicaSet, opts metav1.UpdateOptions) (*v1.VirtualMachineInstanceReplicaSet, error)
-	Delete(ctx context.Context, name string, options metav1.DeleteOptions) error
+	kvcorev1.VirtualMachineInstanceReplicaSetInterface
 	GetScale(replicaSetName string, options metav1.GetOptions) (*autov1.Scale, error)
 	UpdateScale(replicaSetName string, scale *autov1.Scale) (*autov1.Scale, error)
```
Comment (orelmisan) [inline staging/src/kubevirt.io/client-go/kubecli/kubevirt.go:273]: Is there a follow-up planned for adding context to these methods?

**Review by orelmisan — APPROVED** [review]
Thank you @fossedihelm

File: staging/src/kubevirt.io/client-go/kubecli/replicaset.go (hunk 1 above)
Comment (fossedihelm) [reply staging/src/kubevirt.io/client-go/kubecli/replicaset.go:89]: I think it is fine as is for the moment.
We can think about defining a generalized behavior for all the methods in a follow up PR.
WDYT?

File: staging/src/kubevirt.io/client-go/kubecli/kubevirt.go (hunk 2 above)
Comment (fossedihelm) [reply staging/src/kubevirt.io/client-go/kubecli/kubevirt.go:273]: I will add a commit here! Thank you!

File: staging/src/kubevirt.io/client-go/kubecli/replicaset.go (hunk 1 above)
Comment (fossedihelm) [reply staging/src/kubevirt.io/client-go/kubecli/replicaset.go:89]: I cannot see the misalignment you are referring to...
All the methods use the named returning values except `Delete` (that returns directly the generated function output).

Comment (orelmisan) [reply staging/src/kubevirt.io/client-go/kubecli/replicaset.go:89]: It could also be seen under the `ReplicaSetInterface`.

Comment (fossedihelm) [reply staging/src/kubevirt.io/client-go/kubecli/replicaset.go:89]: Consider that in https://github.com/kubevirt/kubevirt/pull/11685 (or similar) interface will be moved in the `vmrs_expansion`. We can make it standard there :) WDYT?

File: staging/src/kubevirt.io/client-go/kubecli/kubevirt.go (hunk 2 above)
Comment (fossedihelm) [reply staging/src/kubevirt.io/client-go/kubecli/kubevirt.go]: Added 2 commits for the context addition to updatescale and getscale.
PTAL

**Review by orelmisan — APPROVED** [review]

**Review by xpivarc — COMMENTED** [review]
/approve

**Pull request merged**